src/lib/process.pl

```diff
@@ -0,0 +1,202 @@
+:- module(process, [
+    process_create/3, 
+    process_id/2,
+    process_release/1, 
+    process_wait/2, 
+    process_wait/3, 
+    process_kill/1
+]).
+
+:- use_module(library(error)).
+:- use_module(library(iso_ext)).
+:- use_module(library(lists), [member/2, maplist/2]).
+:- use_module(library(reif), [tfilter/3]).
+
+
+%% process_create(+Exe, +Args:list, +Options).
+%
+% Create a new process by executing the executable Exe and passing it the Arguments Args.
+%
+% Note: On windows please take note of [windows argument splitting](https://doc.rust-lang.org/std/process/index.html#windows-argument-splitting).
+% 
+% Options is a list consisting of the following options:
+%
+%  * `cwd(+Path)`           Set the processes working directory to `Path`
+%  * `process(-Process)`    `Process` will be assigned a process handle for the spawned process
+%  * `env(+List)`           Don't inherit environment variables and set the variables defined in `List`
+%  * `environment(+List)`   Inherit environment variables and set/override the variables defined in `List`
+%  * `stdin(Spec)`, `stdout(Spec)` or `stderr(Spec)` defines how to redirect the spawned processes io streams
+%
+%  The elements of `List` in `env(List)`/`environment(List)` List must be string pairs using `=/2`.
+%  `env/1` and `environment/1` may not be both specified.
+%
+%  The following stdio `Spec` are available:
+%  
+%  * `std`          inherit the current processes original stdio streams (does currently not account for stdio being changed by `set_input` or `set_output`)
+%  * `file(+Path)`  attach the strea to the file at `Path`
+%  * `null`         discards writes and behaves as eof for read. Equivalent to using `file(/dev/null)`
+%  * `pipe(-Steam)` create a new pipe and assigne one end to the created process and the other end to `Stream`
+%
+% Specifying an option multiple times is an error, when an option is not specified the following defaults apply:
+%
+% - `cwd(".")`
+% - `environment([])`
+% - `stdin(std)`, `stdout(std)`, `stderr(std)`
+%
+process_create(Exe, Args, Options) :- 
+    must_be(chars, Exe),
+    must_be(list, Args),
+    maplist(must_be(chars), Args),
+    must_be(list, Options),
+    must_be_known_options([stdin, stdout, stderr, env, environment, process, cwd], [], Options),
+    check_options(
+        [
+            ([stdin], valid_stdio, stdin(std), stdin(Stdin)),
```

Comment:
Please consider using a dedicated functor for such tuples, such as `option_check(A,B,C,D)`, or maybe you find a better name.

Currently, the representation is:

<pre>
<b>?- write_canonical(([stdin], valid_stdio, stdin(std), stdin(Stdin))).</b>
','('.'(stdin,[]),','(valid_stdio,','(stdin(std),stdin(_104))))
</pre>

Reply:
I have changed it to use `option(A, B, C, D)` .